File: algorithms/649. Dota2 Senate/solution.cc
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> r, d;
        for ( int i = 0; i < senate.size(); ++i )
            'R' == senate[i] ? r.push(i) : d.push(i);
        while ( r.size() && d.size() ) {
            int ri = r.front(), di = d.front();
            r.pop();
            d.pop();
            if ( ri < di )
                r.push(ri + senate.size());
            else
                d.push(di + senate.size());
        }
        return r.size() ? "Radiant" : "Dire";
    }
};

// 81 / 81 test cases passed.
// Status: Accepted
// Runtime: 9 ms
// beats 52.66 %

File: algorithms/649. Dota2 Senate/solution.py
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        r = collections.deque([i for i, c in enumerate(senate) if c == 'R'])
        d = collections.deque([i for i, c in enumerate(senate) if c == 'D'])
        while r and d:
            ri, di = r.popleft(), d.popleft()
            if ri < di:
                r.append(ri + len(senate))
            else:
                d.append(di + len(senate))
        return "Radiant" if r else "Dire"

# 81 / 81 test cases passed.
# Status: Accepted
# Runtime: 105 ms
# beats 66.93 %
